cps/kni.c

```diff
@@ -485,3 +486,314 @@ rm_kni(void)
 		RTE_LOG(ERR, GATEKEEPER, "Error removing %s: %s\n",
 			name, strerror(errno));
 }
+
+/*
+ * Processing ARP and ND packets from the KNI. If responding to
+ * an ARP/ND packet fails, we remove the request from the linked list,
+ * forcing the KNI to issue another resolution request.
+ */
+
+static void
+cps_arp_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused)) enum lls_reply_ty ty, int *pcall_again)
+{
+	struct gatekeeper_if *iface = arg;
+	struct rte_mbuf *created_pkt;
+	struct ether_hdr *eth_hdr;
+	struct arp_hdr *arp_hdr;
+	size_t pkt_size;
+	struct cps_config *cps_conf = get_cps_conf();
+	struct rte_kni *kni;
+	struct rte_mempool *mp;
+	struct cps_request *req;
+	int ret;
+
+	if (pcall_again != NULL)
+		*pcall_again = false;
+
+	mp = cps_conf->net->gatekeeper_pktmbuf_pool[
+		rte_lcore_to_socket_id(cps_conf->lcore_id)];
+	created_pkt = rte_pktmbuf_alloc(mp);
+	if (created_pkt == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: could not allocate an ARP reply on the %s KNI\n",
+			iface->name);
+		goto remove;
+	}
+
+	pkt_size = sizeof(struct ether_hdr) + sizeof(struct arp_hdr);
+	created_pkt->data_len = pkt_size;
+	created_pkt->pkt_len = pkt_size;
+
+	/* Set-up Ethernet header. */
+	eth_hdr = rte_pktmbuf_mtod(created_pkt, struct ether_hdr *);
+	ether_addr_copy(&map->ha, &eth_hdr->s_addr);
+	ether_addr_copy(&iface->eth_addr, &eth_hdr->d_addr);
+	eth_hdr->ether_type = rte_cpu_to_be_16(ETHER_TYPE_ARP);
+
+	/* Set-up ARP header. */
+	arp_hdr = (struct arp_hdr *)(eth_hdr + 1);
+	arp_hdr->arp_hrd = rte_cpu_to_be_16(ARP_HRD_ETHER);
+	arp_hdr->arp_pro = rte_cpu_to_be_16(ETHER_TYPE_IPv4);
+	arp_hdr->arp_hln = ETHER_ADDR_LEN;
+	arp_hdr->arp_pln = sizeof(struct in_addr);
+	arp_hdr->arp_op = rte_cpu_to_be_16(ARP_OP_REPLY);
+	ether_addr_copy(&map->ha, &arp_hdr->arp_data.arp_sha);
+	rte_memcpy(&arp_hdr->arp_data.arp_sip, map->ip_be,
+		sizeof(arp_hdr->arp_data.arp_sip));
+	ether_addr_copy(&iface->eth_addr, &arp_hdr->arp_data.arp_tha);
+	arp_hdr->arp_data.arp_tip = iface->ip4_addr.s_addr;
+
+	if (iface == &cps_conf->net->front)
+		kni = cps_conf->front_kni;
+	else
+		kni = cps_conf->back_kni;
+
+	ret = rte_kni_tx_burst(kni, &created_pkt, 1);
```

Comment:
The LLS block calls `cps_arp_cb()`, so the LLS and CPS blocks concurrently call `rte_kni_tx_burst()`. Does `rte_kni_tx_burst()` support concurrent calls?

(Option 1) If so, adding a comment about the fact that the LLS block calls `cps_arp_cb()` and that `rte_kni_tx_burst()`  is fine with this is essential to reminder future changes of `cps_arp_cb()`. It's also advisable to check that everything else in `cps_arp_cb()` is concurrent safe.

(Option 2) If not, instead of adding a lock mechanism, you can push all the code to the handling of the request CPS_REQ_ARP by the CPS block. That is, the LLS block would only send a request to the mailbox of the CPS block.